test/server/utils/normalizeOptions.test.js

```diff
@@ -85,7 +87,9 @@ describe('normalizeOptions', () => {
       multiCompiler: false,
       options: {
         client: {
-          path: '/custom/path/',
+          webSocketURL: {
+            path: '/custom/path/',
+          },
```

Comment:
We should add one case with `webSocketURL: 'ws://myhost.com:8080/path'`, and check it normalizes host, port and path correctly from the string.

Reply:
Will the webSocketURL string allow me to provide wss:// and actually have the browser request using wss://, or will the normalization turn it back into ws:// if https is false? I'm running a reverse https -> http proxy in front, so the dev server should be https:false, but currently doing so, the websocket logic attempts to use ws:// even though the address bar is an https:// url. Just checking to see if this string webSocketURL option will indeed allow me to now force wss:// to be used for the websocket requests.

Reply:
@bompus you can use `ws` and `wss`, and even `http`/`https` (we normalize them to `ws`/`wss` for backward compatibility). But you should use `webSocketURL` only when you proxy dev server, if you set `https: true` we will use `wss` by default

Reply:
@snitin315 we have this test (in `test/e2e/web-socket-server-and-url.test.js`):
```
client: {
  webSocketURL: `ws://myhost.test:${port2}/foo/test/bar/`,
},
`

Reply:
All good then 👍🏻

Reply:
@alexander-akait I tried this on master after it was merged and browser is still told to request ws:// instead of wss:// , even when wss:// is specified on webSocketURL. Am I missing something in the config, or is this as expected?

My config is:
```
devServer: {
      https: false,
      host: '127.0.0.1',
      port: 8888,
      firewall: false,
      client: {
        webSocketURL: 'wss://myhost.test/ws/'
      }
}
```

Reply:
hm, why you need to change `protocol`?

Reply:
webpack dev server without https and proxy server under https?

Reply:
The dev server is proxied via a protected cloudflared argo tunnel (cloudflare https -> nginx http -> webpack-dev-server http), so the local dev server needs to listen for http instead of https. This is done so dev matches production closer to prevent surprises.

This works in a previous version of v4 (pre-beta3?) and also in v3.

It would be useful if you just detected the window.location.protocol and used that as the default for ws:// or wss://, which is probably what was done previously. With the recent changes, it seems that everything is being toggled off that https:[true/false] option.

I'm running https:true for now to resolve the issue, but it'd be nice to go back to the old config so the reverse proxy matches production and hits all the same cloudflare https -> nginx http logic.

Reply:
@bompus fair enough, my idea - adding `protocol` option to `webSocketURL` (you can use it in `string` notation too), also provide `protocol: 'auto'` to use from `location`, so passing `auto://0.0.0.0:0/` will means all from `location`, what do you think?

Reply:
@alexander-akait that sounds good to me, as I had already expected it to use `ws` or `wss` based on what had been specified in webSocketURL. `auto` works as well, and I assume one would need to use `auto://0.0.0.0:0/wspath/` if they wanted a specific path?

Reply:
>  I assume one would need to use auto://0.0.0.0:0/wspath/ if they wanted a specific path?

Yes, maybe make sense to add `auto` to path too

Reply:
Great. This should solve my use case 👍

Reply:
Put in todo, will solve tomorrow

Reply:
@alexander-akait did you get a chance to start on this yet, or still pending? Just want to make sure it doesn't get buried.

Reply:
in my todo on tomorrow, don't worry :+1: